### DISCUSSION TITLE
Best way to handle connection object in NATS?

### DISCUSSION THREAD

**Discussion by YashasAnand** [nats-io/nats.net#654] [Q&A] [answered]
Is it better to create a long lived connection instance to the NATS broker, or ignore creating the connection as internally publish method is creating connection. 

Also in case of consumer, does the library internally handle connection to broker if its not connected already?

Am i getting a performance improvement if i create a long lived connection to broker or is it actually degrading my performance by consuming memory ?

**Comment by mtmk** [comment]
Yes, typically you should create a connection once as a singleton and use the same object until your application exists. Library internally maintains the connection, reconnects after connection is lost and maintains subscriptions and consume calls.

**Reply by YashasAnand** [reply]
@mtmk , thank you for your response, I have observed that if connection is not created as a singleton, any messages published to stream will internally open a connection, publish the messages & immediately close. 

As per your above comment, you say the client tries to reconnect the same connection when new messages published.. 
With respect to performance, is it is better to have a singleton instance than use the internal connect-reconnect mechanism?

**Reply by mtmk** [reply, marked as answer]
that's correct. you'd want the connection to be a singleton so it's established and kept alive internally (by reconnecting when needed)

**Reply by YashasAnand** [reply]
Thanks again @mtmk !!

**Comment by YashasAnand** [comment]
@mtmk hope you're doing well, just a quick question, since its suggested to use single connection per service model, does that mean its also preferred to have single publisher per service?

Also can you pls suggest what are the pitfalls of having many connections per service or also many producers per subject?

what is the ideal design should we 
1) have many connections & many publishers?
2) 1 connection many publishers objects
3) 1 connection 1 publisher object


thank you for your time.

**Reply by mtmk** [reply]
thanks @YashasAnand I'm very well hope you're well too. NATS protocol is designed to handle multiple publishers and subscribers per connection. it depends on the expected message volume, type of application, host spec/number of nodes, server cluster, network etc. how many you can have per connection. for example you can have many more Core NATS publishers than JetStream publishers since JetStream is more chatty in comparison. I suggest to test your design if you're expecting to handle loads of messages. but for small volumes you should be fine. you can use `nats bench` cli utility and you should prototype and test your NATS .NET application as much as possible. hope it helps.

**Reply by YashasAnand** [reply]
@mtmk thank you for the quick response, im doing good as well. We have more jetstream topics & publishes to jetstream are more. corenats publishes are comparitively lesser.

My only concern is i have a singleton nats class which handles both connection & publish operations. If we end up getting exception on any produce will it end up stalling other publishes? 

we have around 60 subjects in total in micoservice set up per microservice around 10.

pls could you let me know how using natsbench i can test my application, in my application some rest endpoint trigger some produce event.

**Reply by mtmk** [reply]
> My only concern is i have a singleton nats class which handles both connection & publish operations. If we end up getting exception on any produce will it end up stalling other publishes?

It should be fine. NATS .NET is fully async so we rely on the threadpool. as long as you keep everything async/await style and not block any threadpool threads you should be fine as well. TCP connection limits apply of course but NATS protocol is asynchronous and can handle multiple publishers without stalling. 
 
> pls could you let me know how using natsbench i can test my application, in my application some rest endpoint trigger some produce event.

to my understanding, you can use `nats bench` to get an idea of 'theoretical' limits of your NATS setup or building blocks of a particular scenario. it's part of the [NATS cli](https://github.com/nats-io/natscli) after installing it just run `nats bench` to see a few examples of how to use it. cc @jnmoyne

**Reply by YashasAnand** [reply]
@mtmk thank you again for your response , i have a nats class library which is fully async

but my applications which are kind of legacy codes are not fully async so i am doing this while producing can you pls let me know if this can have issues?

   NatsService.NatsInstance.ProduceAsync(topic,message).Wait(); //at application layer
   
   This invokes class library method as below
```
   public async Task ProduceAsync(string topic, Message<string, string> message)
    {
	 NatsMessageModel natsMessage = new()
            {
                Key = message.Key,
                Value= message.Value
	    }
            
                await _natsJSContext.PublishAsync(js_topic,natsMessage,serializer:natsMesgSerializer));
   }
```

Will this have any issues? also do you suggest i make this a fire& forget type like?
_ = NatsService.NatsInstance.ProduceAsync(topic,message);

**Reply by mtmk** [reply]
difficult to make suggestions without knowing much about the application and I don't claim to be an expert on this. i'd suggest to research how to mix legacy/non-async and async code and pick one suits you best. couple of things I can say is make sure you don't do fire-and-forget without some form of control (e.g. using semaphores) otherwise you can overwhelm your application and server with too many request. best to stick to awaiting publish calls. if throughput is an issue you can use concurrent publishing too. also try to keep blocking calls off the threadpool unless they are truly cpu workloads. and perhaps look into using channels. they do have async and non async calls which might help. good luck!

**Reply by YashasAnand** [reply]
thank you for the valuable suggestions @mtmk
